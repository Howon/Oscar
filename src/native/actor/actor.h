#include <thread>

class actor {
public:
    actor()

};
