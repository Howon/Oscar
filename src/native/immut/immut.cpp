#include "list.h"
#include "map.h"
#include "set.h"

#include <vector>
#include <string>

template <typename T, typename F>
void ForEach(const F f, const immut::list<T> &t)
{
    t.forEach(f);
}

template <typename T, typename F>
void ForEach(const F f, const immut::set<T> &t)
{
    t.forEach(f);
}

template <typename K, typename V, typename F>
void ForEach(const F f, const immut::map<K, V> &t)
{
    t.forEach(f);
}

template<class T, class U, class F>
U FoldLeft(F f, U acc, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(U, T)>>::value,
        "FoldLeft requires a function type U(U, T)");

    ForEach([&f, &acc](const T &v){
        acc = f(acc, v);
    }, lst);

    return acc;
}

template<class T>
immut::list<T> Reverse(immut::list<T> const &lst)
{
    return FoldLeft([](immut::list<T> const &acc, T v) {
        return immut::list<T>(v, acc);
    }, immut::list<T>(), lst);
}

template <typename T>
size_t Size(immut::list<T> l)
{
    return FoldLeft([](size_t acc, T t) -> size_t {
        return acc + 1;
    }, 0, l);
}

template <typename T>
size_t Size(immut::set<T> s)
{
    size_t size = 0;

    ForEach([&size](T t) {
        size++;
    }, s);

    return size;
}

template <typename K, typename V>
size_t Size(immut::map<K, V> m)
{
    size_t size = 0;

    ForEach([&size](K k, V v) {
        size++;
    }, m);

    return size;
}


template <typename T>
immut::set<T> Put(T x, immut::set<T> s)
{
    return s.inserted(x);
}

template <typename K, typename V>
immut::map<K, V> Put(K k, V v, immut::map<K, V> m)
{
    return m.inserted(k, v);
}

template <typename T, typename F>
immut::list<T> Filter(F f, immut::list<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::list<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = res.push_front(v);
    }, t);

    return Reverse<T>(res);
}

template <typename T, typename F>
immut::set<T> Filter(F f, immut::set<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::set<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = Put(v, res.inserted(v));
    }, t);

    return res;
}

template <typename K, typename V, typename F>
immut::map<K, V> Filter(F f, immut::map<K, V> t)
{
    static_assert(is_convertible<F, function<bool(K, V)>>::value,
        "Filter requires a function type bool(K, V)");

    immut::map<K, V> res;

    ForEach([&f, &res, &t](const K &k, const V &v){
        if (f(k, v))
            res = Put(k, v, res);
    }, t);

    return res;
}

template <typename T>
bool Contains(T t, immut::list<T> l)
{
    return l.contains(t);
}

template <typename T>
bool Contains(T t, immut::set<T> s)
{
    return s.contains(t);
}

template <typename K, typename V>
bool Contains(K k, immut::map<K, V> m)
{
    return m.contains(k);
}

template<class T, class F>
auto Map(F f, immut::list<T> t) -> immut::list<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Map requires a function type U(T)");

    immut::list<U> res;

    t.forEach([&f, &res](const T &v) { res = res.push_front(f(v)); });

    return Reverse<U>(res);
}

template<class T, class F>
auto Map(F f, immut::set<T> t) -> immut::set<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Met requires a function type U(T)");

    immut::set<U> res;

    t.forEach([&f, &res](const T &v) { res = Put<U>(f(v), res); });

    return res;
}

template <typename K, typename V, typename F>
auto Map(F f, immut::map<K, V> t) ->
    immut::map<K, decltype(f(t.rootKey(), t.rootValue()))>
{
    using U = decltype(f(t.rootKey(), t.rootValue()));

    static_assert(is_convertible<F, function<U(K, V)>>::value,
        "Map requires a function type U(K, V)");

    immut::map<K, U> res;

    t.forEach([&f, &res](K const &k, V const &v) {
        res = Put(k, f(k, v), res);
    });

    return res;
}

template <typename T, typename U, typename F>
U Reduce(F f, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(T, T)>>::value,
        "Reduce requires a function type U(T, T)");
    static_assert(!lst.isEmpty(),
        "Reduce requires a nonempty collection");

    auto next = lst.popped_front;

    return FoldLeft(f, f(lst.front(), next.front()), next.popped_front());
}

template <typename T>
immut::list<T> MergeFront(immut::list<T> const &a, immut::list<T> const &b)
{
    if (a.isEmpty())
        return b;

    return immut::list<T>(a.front(), MergeFront(a.pop_front(), b));
}

template <typename T>
immut::list<T> MergeBack(immut::list<T> const &a, immut::list<T> const &b)
{
    return MergeFront(b, a);
}

template <typename T>
immut::list<T> PopFront(immut::list<T> const l)
{
    return l.pop_front();
}

template <typename T>
immut::list<T> PopBack(immut::list<T> const l)
{
    size_t size = Size(l) - 1;

    return l.removeAt(0, size);
}

template <typename T>
immut::list<T> PushFront(T t, immut::list<T> const l)
{
    return l.push_front(t);
}

template <typename T>
immut::list<T> PushBack(T t, immut::list<T> const l)
{
    auto r1 = Reverse(l);
    auto r2 = r1.push_front(t);

    return Reverse(r2);
}

template <typename T>
immut::set<T> Union(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res = a;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Diff(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Intersection(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
bool SubSet(immut::set<T> const & a, immut::set<T> const & b)
{
    bool is_subset = true;

    ForEach([&is_subset, &b](const T &v) {
        if (!b.contains(v))
            is_subset = false;
    }, a);

    return is_subset;
}

/*
void test_list(void)
{
    immut::list<int> lst = { 1, 2, 3, 4, 5, 6, 7, 8, 9};

    cout << lst << endl;

    auto mapped = Map([](int x) -> immut::list<double> {
        return immut::list<double>{ x + 0.5 };
    }, lst);

    cout << mapped << endl;

    auto fold = FoldLeft([](double acc, int x) -> double {
        return acc * x;
    }, 10.0, lst);

    cout << fold << endl;

    auto filter = Filter([](int x) -> bool {
        return x % 2 == 0;
    }, lst);

    cout << filter << endl;

    auto mergeF = MergeFront(lst, filter);
    auto mergeB = MergeBack(lst, filter);

    cout << mergeF << endl;
    cout << mergeB << endl;

    auto pushF = PushFront(0, lst);
    auto pushB = PushBack(0, lst);

    auto popF = PopFront(lst);
    auto popB = PopBack(lst);

    cout << pushF << endl;
    cout << pushB << endl;

    cout << popF << endl;
    cout << popB << endl;

    cout << Contains(1, lst) << endl;
    cout << Contains(11, lst) << endl;
    cout << Size(lst) << endl;
    cout << lst << endl << endl;
}

void test_set(void)
{
    immut::set<int> set = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    immut::set<int> set2 = {4, 5, 6, 7, 8, 9, 9, 10, 11, 12};
    immut::set<int> set3 = {1, 2, 3, 4, 5, 6, 9};

    cout << set << endl;
    cout << set2 << endl;

    auto mapped = Map([](int x) -> immut::list<double> {
        return immut::list<double>{ x + 0.5 };
    }, set);

    cout << mapped << endl;

    auto filter = Filter([](int x) -> bool {
        return x % 2 == 0;
    }, set);

    cout << filter << endl;

    auto uni = Union(set, set2);
    auto diff = Diff(set, set2);
    auto intersect = Intersection(set, set2);

    cout << uni << endl;
    cout << diff << endl;
    cout << intersect << endl;
    cout << SubSet(set, set2) << endl;
    cout << SubSet(set, set3) << endl;
    cout << SubSet(set3, set) << endl;
    cout << Put(12, set) << endl;
    cout << Size(set) << endl;
    cout << set << endl << endl;
}

void test_map(void)
{
    immut::map<string, int> map = { {"a", 1}, {"b", 2}, {"c", 3}, {"d", 4} };

    auto mapped = Map([](string s, int x) -> immut::list<double> {
        return immut::list<double> { x + 0.5, 4.5 };
    }, map);

    cout << mapped << endl;

    auto filter = Filter([](string s, int x) -> bool {
        return x % 2 == 0;
    }, map);

    cout << filter << endl;
    cout << Contains(string("b"), map) << endl;
    cout << Contains(string("z"), map) << endl;
    cout << Put(string("h"), 12, map) << endl;
    cout << Size(map) << endl;
    cout << map << endl;
}

int main(void)
{
    test_list();
    test_set();
    test_map();

    return 0;
}
*/
