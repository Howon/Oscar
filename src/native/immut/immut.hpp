#ifndef __IMMUT_HPP__
#define __IMMUT_HPP__

#include "list.hpp"
#include "map.hpp"
#include "set.hpp"

#include <iomanip>
#include <string>

template <typename T, typename F>
void ForEach(const F f, const immut::list<T> &t)
{
    t.forEach(f);
}

template <typename T, typename F>
void ForEach(const F f, const immut::set<T> &t)
{
    t.forEach(f);
}

template <typename K, typename V, typename F>
void ForEach(const F f, const immut::map<K, V> &t)
{
    t.forEach(f);
}

template<class T, class U, class F>
U FoldLeft(F f, U acc, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(U, T)>>::value,
        "FoldLeft requires a function type U(U, T)");

    ForEach([&f, &acc](const T &v){
        acc = f(acc, v);
    }, lst);

    return acc;
}

template<class T>
immut::list<T> Reverse(immut::list<T> const &lst)
{
    return FoldLeft([](immut::list<T> const &acc, T v) {
        return immut::list<T>(v, acc);
    }, immut::list<T>(), lst);
}

int Size(string s)
{
    return s.size();
}

template <typename T>
int Size(immut::list<T> l)
{
    return FoldLeft([](size_t acc, T t) -> size_t {
        return acc + 1;
    }, 0, l);
}

template <typename T>
int Size(immut::set<T> s)
{
    size_t size = 0;

    ForEach([&size](T t) {
        size++;
    }, s);

    return size;
}

template <typename K, typename V>
int Size(immut::map<K, V> m)
{
    size_t size = 0;

    ForEach([&size](K k, V v) {
        size++;
    }, m);

    return size;
}


template <typename T>
immut::set<T> Put(T x, immut::set<T> s)
{
    return s.inserted(x);
}

template <typename K, typename V>
immut::map<K, V> Put(K k, V v, immut::map<K, V> m)
{
    return m.inserted(k, v);
}

template <typename T, typename F>
immut::list<T> Filter(F f, immut::list<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::list<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = res.push_front(v);
    }, t);

    return Reverse<T>(res);
}

template <typename T, typename F>
immut::set<T> Filter(F f, immut::set<T> t)
{
    static_assert(is_convertible<F, function<bool(T)>>::value,
        "Filter requires a function type bool(T)");

    immut::set<T> res;

    ForEach([&f, &res, &t](T const &v){
        if (f(v))
            res = Put(v, res.inserted(v));
    }, t);

    return res;
}

template <typename K, typename V, typename F>
immut::map<K, V> Filter(F f, immut::map<K, V> t)
{
    static_assert(is_convertible<F, function<bool(K, V)>>::value,
        "Filter requires a function type bool(K, V)");

    immut::map<K, V> res;

    ForEach([&f, &res, &t](const K &k, const V &v){
        if (f(k, v))
            res = Put(k, v, res);
    }, t);

    return res;
}

template <typename T>
bool Contains(T t, immut::list<T> l)
{
    return l.contains(t);
}

template <typename T>
bool Contains(T t, immut::set<T> s)
{
    return s.contains(t);
}

template <typename K, typename V>
bool Contains(K k, immut::map<K, V> m)
{
    return m.contains(k);
}

template<class T, class F>
auto Map(F f, immut::list<T> t) -> immut::list<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Map requires a function type U(T)");

    immut::list<U> res;

    t.forEach([&f, &res](const T &v) { res = res.push_front(f(v)); });

    return Reverse<U>(res);
}

template<class T, class F>
auto Map(F f, immut::set<T> t) -> immut::set<decltype(f(t.front()))>
{
    using U = decltype(f(t.front()));

    static_assert(is_convertible<F, function<U(T)>>::value,
        "Met requires a function type U(T)");

    immut::set<U> res;

    t.forEach([&f, &res](const T &v) { res = Put<U>(f(v), res); });

    return res;
}

template <typename K, typename V, typename F>
auto Map(F f, immut::map<K, V> t) ->
    immut::map<K, decltype(f(t.rootKey(), t.rootValue()))>
{
    using U = decltype(f(t.rootKey(), t.rootValue()));

    static_assert(is_convertible<F, function<U(K, V)>>::value,
        "Map requires a function type U(K, V)");

    immut::map<K, U> res;

    t.forEach([&f, &res](K const &k, V const &v) {
        res = Put(k, f(k, v), res);
    });

    return res;
}

template <typename T, typename U, typename F>
U Reduce(F f, immut::list<T> lst)
{
    static_assert(is_convertible<F, function<U(T, T)>>::value,
        "Reduce requires a function type U(T, T)");
    static_assert(!lst.isEmpty(),
        "Reduce requires a nonempty collection");

    auto next = lst.popped_front;

    return FoldLeft(f, f(lst.front(), next.front()), next.popped_front());
}

template <typename T>
immut::list<T> MergeFront(immut::list<T> const &a, immut::list<T> const &b)
{
    if (a.isEmpty())
        return b;

    return immut::list<T>(a.front(), MergeFront(a.pop_front(), b));
}

template <typename T>
immut::list<T> MergeBack(immut::list<T> const &a, immut::list<T> const &b)
{
    return MergeFront(b, a);
}

template <typename T>
immut::list<T> PopFront(immut::list<T> const l)
{
    return l.pop_front();
}

template <typename T>
immut::list<T> PopBack(immut::list<T> const l)
{
    size_t size = Size(l) - 1;

    return l.removeAt(0, size);
}

template <typename T>
immut::list<T> Prepend(T t, immut::list<T> const l)
{
    return l.push_front(t);
}

template <typename T>
immut::list<T> Append(T t, immut::list<T> const l)
{
    auto r1 = Reverse(l);
    auto r2 = r1.push_front(t);

    return Reverse(r2);
}

template <typename T>
immut::set<T> Union(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res = a;

    ForEach([&res, &a](const T &v){
        if (!a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
immut::set<T> Diff(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &b](const T &v){
        if (!b.contains(v))
            res = res.inserted(v);
    }, a);

    return res;
}

template<class T>
immut::set<T> Intersection(immut::set<T> const & a, immut::set<T> const & b)
{
    immut::set<T> res;

    ForEach([&res, &a](const T &v){
        if (a.contains(v))
            res = res.inserted(v);
    }, b);

    return res;
}

template<class T>
bool SubSet(immut::set<T> const & a, immut::set<T> const & b)
{
    bool is_subset = true;

    ForEach([&is_subset, &b](const T &v) {
        if (!b.contains(v))
            is_subset = false;
    }, a);

    return is_subset;
}

int AsInt(double d) { return int(d); }

double AsDouble(int i) { return double(i); }

std::string AsString(int i) { return std::to_string(i); }

std::string AsString(char c) { return std::to_string(c); }

std::string AsString(bool b) { return std::to_string(b); }

std::string AsString(double d) { return std::to_string(d); }

template <typename T>
std::string AsString(immut::list<T> l) { return l.str(); }

template <typename T>
std::string AsString(immut::set<T> s) { return s.str(); }

template <typename K, typename V>
std::string AsString(immut::map<K, V> m) { return m.str(); }

void Println(int i) { std::cout << i << std::endl; }

void Println(char c) { std::cout << c << std::endl; }

void Println(bool b) {
    std::cout << (b == true ? "true" : "false") << std::endl;
}

void Println(double d) {
    std::cout << fixed << setprecision(10);
    std::cout << d << std::endl;
}

void Println(string s) { std::cout << s << std::endl; }

template <typename T>
void Println(immut::list<T> l) { std::cout << l << std::endl; }

template <typename T>
void Println(immut::set<T> s) { std::cout << s << std::endl; }

template <typename K, typename V>
void Println(immut::map<K, V> m) { std::cout << m << std::endl; }

#endif
