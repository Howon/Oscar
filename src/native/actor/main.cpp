#include "actor.hpp"
using namespace std;

// todo: keep a global map of <address, Actor> for communication


int main() {
    Actor a;

    HelloMessage hm(1, "hello");
    ByeMessage bm("bye");

    a.receive(hm);
    a.receive(bm);

    return 0;
}