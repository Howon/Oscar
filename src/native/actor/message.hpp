#ifndef __MESSAGE__
#define __MESSAGE__

#include <string>
#include <tuple>
#include <vector>

#include "actor.hpp"

using namespace std;

enum MessageType { None, Some };

class Actor;

class Message {
// todo: make private
public:
    string name;
    MessageType type;
    Actor* sender;

public:
    Message(const string& name="", MessageType type=None, Actor* const sender=NULL)
            : name(name), type(type), sender(sender)
    { }

    bool operator==(const Message& rhs) const {
        if (this == &rhs)
            return true;

        return (this->type == rhs.type &&
                this->name == rhs.name);
    }

    bool operator!=(const Message& rhs) const {
        return !(*this == rhs);
    }
};


//class EmptyMessage : public Message {
//    string s;
//
//public:
//    EmptyMessage() : s("")
//    { }
//
//    tuple<string> get() { return make_tuple(s); }
//};

static Message EMPTY_MESSAGE;


class HelloMessage : public Message {
    int i;
    string s;

public:
    HelloMessage(int x, string y) : Message("hello", Some) {
        x = i;
        s = y;
    }

    tuple<int, string> get() { return make_tuple(i, s); }
};


class ByeMessage : public Message {
    string s;

public:
    ByeMessage(string y) : Message::Message("bye", Some) {
        s = y;
    }

    tuple<string> get() { return make_tuple(s); }
};


#endif //__MESSAGE__
