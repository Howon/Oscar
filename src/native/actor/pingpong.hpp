#ifndef __PINGPONG__
#define __PINGPONG__

#include "actor.hpp"
#include "../immut/immut.hpp"


class Pong : public Actor {
    queue<StopMessage *> stopQueue;
    queue<PingMessage *> pingQueue;

    void consume() {
        unique_lock<mutex> lck(mx);

        while (!this->tFinished) {
            while (stopQueue.empty() && pingQueue.empty()) {
                // finished running, terminate thread
                if (tFinished)
                    return;

                // else, wait for messages to arrive
                cv.wait(lck);
                goto loop;
            }

            // process current message
            if (!stopQueue.empty()) {
                auto msg = stopQueue.front();
                stopQueue.pop();
                respond(msg);
            } else if (!pingQueue.empty()) {
                auto msg = pingQueue.front();
                pingQueue.pop();
                respond(msg);
                goto loop;
            }
            loop: ;
        }
    }

    void respond(StopMessage *msg) {
        delete msg;
        printf("pong stopped\n\n");
        Die();
    }

    void respond(PingMessage *msg) {
        printf("  pong\n");
        msg->sender->receive(new PongMessage(this));

        delete msg;
    }

public:
    Pong() {
        t = thread([=] { consume(); });
    }

    virtual ~Pong() {
        t.join();
    }

    void receive(Message* const msg) {
        if (PingMessage* pm = dynamic_cast<PingMessage *>(msg)) {
            unique_lock<mutex> lck(mx);
            pingQueue.push(pm);
        }
        else if (StopMessage *pm = dynamic_cast<StopMessage*>(msg)) {
            unique_lock<mutex> lck(mx);
            stopQueue.push(pm);
        }

        cv.notify_one();
    }
};

class Ping : public Actor {
    Pong* pong;

    queue<StartMessage *> startQueue;
    queue<PongMessage *> pongQueue;

    int count = 0;
    int maxTurns;

        void consume()
    {
        unique_lock<mutex> lck(mx);
        while (!this->tFinished) {
            while (pongQueue.empty() && startQueue.empty()) {
                if (tFinished)
                    return;
                cv.wait(lck);
            }
            if (!pongQueue.empty()) {
                auto msg = pongQueue.front();
                pongQueue.pop();
                respond(msg);
                goto loop;
            }
            if (!startQueue.empty()) {
                auto msg = startQueue.front();
                startQueue.pop();
                respond(msg);
                goto loop;
            }
        loop:;
        }
    }

    void incrementAndPrint() {
        ++count;
        printf("ping\n");
    }

    void respond(StartMessage *msg) {
        delete msg;
        incrementAndPrint();
        pong->receive(new PingMessage(this));
    }


    void respond(PongMessage* msg)
    {
        delete msg;
        incrementAndPrint();
        if (count > maxTurns) {
            Println(std::string("ping stopped"));
            pong->receive(new StopMessage(this));

            Die();
        }
        else {
            pong->receive(new PingMessage(this));
        }
    }

public:
    Ping(Pong* pong, int maxTurns=99) : Actor() {
        this->pong = pong;
        this->maxTurns = maxTurns;

        t = thread([=] { consume(); });
    }

    virtual ~Ping() {
        t.join();
    }

    void receive(Message* const msg) {
       if (PongMessage* pm = dynamic_cast<PongMessage*>(msg)) {
            unique_lock<mutex> lck(mx);
            pongQueue.push(pm);
            goto notify;
        }
        if (StartMessage* pm = dynamic_cast<StartMessage*>(msg)) {
            unique_lock<mutex> lck(mx);
            startQueue.push(pm);
            goto notify;
        }
    notify:
        cv.notify_one();
    }
};


#endif //__PINGPONG__
