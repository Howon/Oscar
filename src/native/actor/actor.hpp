#ifndef __ACTOR__
#define __ACTOR__

#include "iostream"
#include <cstdio>
#include <queue>
#include <string>
#include <unordered_map>

#include <condition_variable>
#include <mutex>
#include <thread>

#include "message.hpp"

using namespace std;

class Actor {
public:
    std::thread t;
    condition_variable cv;
    mutex mx;
    bool tFinished;

public:
    Actor() : tFinished(false) { }

    ~Actor() { }

    virtual void receive(Message* const msg) = 0;

    void Die() { this->tFinished = true; }
};

class Monitor : public Actor {
    bool end;
    atomic<int> actor_counter;

    queue<SpawnMessage *> spawnQueue;
    queue<DeleteMessage *> deleteQueue;

    void consume() {
        unique_lock<mutex> lck(mx);

        while (!this->tFinished) {
            while (spawnQueue.empty() && deleteQueue.empty()) {
                if (tFinished)
                    return;
                cv.wait(lck);
            }

            if (!spawnQueue.empty()) {
                auto msg = spawnQueue.front();
                spawnQueue.pop();
                respond(msg);
            } else if (!deleteQueue.empty()) {
                auto msg = deleteQueue.front();
                deleteQueue.pop();
                respond(msg);
            }
        }
    }

    void respond(SpawnMessage *msg) {
        actor_counter++;
        end = false;

        delete msg;
    }

    void respond(DeleteMessage *msg) {
        actor_counter--;

        delete msg;

        if (actor_counter.load() == 0 && spawnQueue.empty() &&
            deleteQueue.empty())
            this->end = true;
    }

public:
    Monitor() {
        this->end = true;
        actor_counter = 0;
        t = thread([=] { consume(); });
    }

    virtual ~Monitor() {
        this->end = true;

        t.join();
    }

    bool is_exitable() { return this->end; }

    void receive(Message* const msg) {
        if (SpawnMessage* pm = dynamic_cast<SpawnMessage *>(msg)) {
            unique_lock<mutex> lck(mx);
            spawnQueue.push(pm);
        } else if (DeleteMessage *pm = dynamic_cast<DeleteMessage *>(msg)) {
            unique_lock<mutex> lck(mx);
            deleteQueue.push(pm);
        }

        cv.notify_one();
    }
};

#endif  // __ACTOR__
