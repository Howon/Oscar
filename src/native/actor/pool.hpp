#ifndef __POOL__
#define __POOL__

#include <vector>
#include <queue>

#include <condition_variable>
#include <mutex>
#include <thread>

#include "actor.hpp"
#include "pingpong.hpp"
using namespace std;

class Pool {
    vector<Actor *> workers;
    queue<Message *> tasks;
    atomic<int> counter;
    condition_variable cv;
    mutex mx;
    bool tFinished;

    void consume() {
        unique_lock<mutex> lck(mx);

        while (true) {
            while (tasks.empty())
                cv.wait(lck);

            int i = (counter++) % workers.size();
            Message* msg = tasks.front();
            tasks.pop();

            workers[i]->receive(msg);
        }
    }

public:
//    Pool(T *actor, int cap = 3) : tFinished(false) {
    Pool(Actor *actor, int cap = 3) : tFinished(false) {
        counter = 0;
        for (int i = 0; i < cap; ++i) {
            workers.push_back(actor);
        }
//                    [this] {
//                        while (true) {
//                            std::function<void()> task;
//
//                            {
//                                std::unique_lock<std::mutex> lock(mx);
//                                cv.wait(lock,
//                                        [this] { return tFinished || !tasks.empty(); }
//                                );
//                                if (tFinished && tasks.empty())
//                                    return;
//
//                                // todo: compilation breaks here b/c Message is incompatible with function<void()>
//                                task = tasks.front();
//                                tasks.pop();
//                            }
//
//                            task();
//                        }
//                    }
//            );
//        }
    }

    ~Pool() {
        {
            unique_lock<mutex> lck(mx);
            tFinished = true;
        }

        cv.notify_all();
        for (Actor* a : workers)
            delete a;
    }

    // todo: this is equivalent of ThreadPool::enqueue
    // todo: needs love!!!
    void receive(Message* msg) {
        {
            unique_lock<mutex> lck(mx);
            tasks.push(msg);
        }

        cv.notify_one();
    }

};


#endif //__POOL__
