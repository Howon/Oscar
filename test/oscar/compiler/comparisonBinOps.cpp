#include <oscar/actor.hpp>
#include <oscar/immut.hpp>
class m_die : public Message {
private:

public:
m_die (Actor *sender) : Message("die", sender)
{

}
tuple<> get() {
return make_tuple();
}
};

class m_print : public Message {
private:
std::string msg;

public:
m_print (std::string msg,Actor *sender) : Message("print", sender)
{
this->msg = msg;

}
tuple<std::string> get() {
return make_tuple(this->msg);
}
};


class a_Printer : public Actor {
private:
std::queue<m_die *> dieQueue;
std::queue<m_print *> printQueue;




public:
a_Printer () : Actor()
{

t = thread([=] { consume(); });
}
virtual ~a_Printer() {
Die();
}

void Die() {
this->tFinished = true;
t.join();
}
virtual void receive(Message *msg) {
if (m_die *pm = dynamic_cast<m_die *>(msg)) {
unique_lock<mutex> lck(mx);
dieQueue.push(pm);
goto notify;
}
if (m_print *pm = dynamic_cast<m_print *>(msg)) {
unique_lock<mutex> lck(mx);
printQueue.push(pm);
goto notify;
}
notify:
cv.notify_one();
}
void respond(m_print *theMsgThatWasReceived) {
auto msg = get<0>(theMsgThatWasReceived->get());

Println(msg);
delete theMsgThatWasReceived;
}

void respond(m_die *theMsgThatWasReceived) {


delete theMsgThatWasReceived;
Die();
}
void consume() {
unique_lock<mutex> lck(mx);
while (!this->tFinished){
while (dieQueue.empty()&&printQueue.empty()){
if (tFinished)
return;
cv.wait(lck);
}
if (!dieQueue.empty()) {
auto msg = dieQueue.front();
dieQueue.pop();
respond(msg);
goto loop;
}
if (!printQueue.empty()) {
auto msg = printQueue.front();
printQueue.pop();
respond(msg);
goto loop;
}
loop: ;
}
}};


int main () {
Println(true);
Println(false);
Println(true);
Println(false);
Println(true);
Println(false);
Println(true);
Println(false);
Println(true);
Println(false);
Println(true);
Println(false);
return 0;
}

