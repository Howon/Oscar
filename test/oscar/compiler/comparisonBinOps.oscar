def main() => unit = {
  println(5 == 5);
  println(5 != 5);
  println(4 < 5);
  println(5 <= 5);
  println(3 > 5);
  println(5 >= 5);
}