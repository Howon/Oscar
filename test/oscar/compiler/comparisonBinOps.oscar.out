true
false
true
true
false
true
